### PR TITLE
Face dataset

### PR DESCRIPTION
**Pull request by jin-s13**

### CODE REVIEW

File: mmpose/core/evaluation/top_down_eval.py [hunk 1]
```diff
@@ -202,6 +202,30 @@ def keypoint_auc(pred, gt, mask, normalize, num_step=20):
     return auc
 
 
+def keypoint_nme(pred, gt, mask, normalize):
+    """Calculate the end-point error.
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:206]: what does `nme` mean

File: mmpose/datasets/datasets/face/face_300w_dataset.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        self.ann_info['flip_pairs'] = [[1, 17], [2, 16], [3, 15], [4, 14],
+                                       [5, 13], [6, 12], [7, 11], [8, 10],
+                                       [18, 27], [19, 26], [20, 25], [21, 24],
+                                       [22, 23], [32, 36], [33, 35], [37, 46],
+                                       [38, 45], [39, 44], [40, 43], [41, 48],
+                                       [42, 47], [49, 55], [50, 54], [51, 53],
+                                       [62, 64], [61, 65], [68, 66], [59, 57],
+                                       [60, 56]]
+
+        self.dataset_name = '300w'
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_300w_dataset.py:56]: `300w`, so concise a name

File: mmpose/core/evaluation/top_down_eval.py (hunk 1 above)
Comment (jin-s13) [reply mmpose/core/evaluation/top_down_eval.py:206]: `normalized mean error (NME)` I will add this information.

File: mmpose/datasets/datasets/face/face_300w_dataset.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/face/face_300w_dataset.py:56]: 300W is the standard abbreviation for this dataset : )

File: mmpose/core/evaluation/top_down_eval.py [hunk 3]
```diff
@@ -202,6 +202,30 @@ def keypoint_auc(pred, gt, mask, normalize, num_step=20):
...
+        batch_size: N
+        num_keypoints: K
+
+    Args:
+        pred (np.ndarray[N, K, 2]): Predicted keypoint location.
+        gt (np.ndarray[N, K, 2]): Groundtruth keypoint location.
+        mask (np.ndarray[N, K]): Visibility of the target. False for invisible
+            joints, and True for visible. Invisible joints will be ignored for
+            accuracy calculation.
+        normalize (np.ndarray[N, 2]): Normalization factor.
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:218]: `normalize` -> `normalize_factor`
`normalize` looks like a switch

File: mmpose/core/evaluation/top_down_eval.py [hunk 4]
```diff
@@ -202,6 +202,30 @@ def keypoint_auc(pred, gt, mask, normalize, num_step=20):
...
+
+    Returns:
+        float: normalized mean error
+    """
+    distances = _calc_distances(pred, gt, mask, normalize)
+    distance_valid = distances[distances != -1]
+    valid_num = len(distance_valid)
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:225]: what if num valid = 0?

File: mmpose/datasets/datasets/face/face_300w_dataset.py (hunk 2 above)
Comment (innerlee) [reply mmpose/datasets/datasets/face/face_300w_dataset.py:56]: okay, `w` means `wild` and not `wan`

Comment (jin-s13) [reply mmpose/datasets/datasets/face/face_300w_dataset.py:56]: 300 images in the wild   hhhh

File: configs/face/hrnet/300w/hrnet_w18_300w_256x256.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    dataset_joints=68,
+    dataset_channel=[
+        list(range(68)),
+    ],
+    inference_channel=list(range(68)))
+
+# model settings
+model = dict(
+    type='TopDown',
+    pretrained='msra/hrnetv2_w18',
```
Comment (innerlee) [inline configs/face/hrnet/300w/hrnet_w18_300w_256x256.py:35]: replace with url?

File: mmpose/core/evaluation/top_down_eval.py [hunk 6]
```diff
@@ -202,6 +202,33 @@ def keypoint_auc(pred, gt, mask, normalize, num_step=20):
...
+            joints, and True for visible. Invisible joints will be ignored for
+            accuracy calculation.
+        normalize_factor (np.ndarray[N, 2]): Normalization factor.
+
+    Returns:
+        float: normalized mean error
+    """
+    distances = _calc_distances(pred, gt, mask, normalize_factor)
+    distance_valid = distances[distances != -1]
+    valid_num = len(distance_valid)
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:225]: `return distance_valid.sum() / max(1, len(distance_valid))`

File: mmpose/datasets/datasets/face/face_300w_dataset.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                joints_3d = np.zeros((num_joints, 3), dtype=np.float32)
+                joints_3d_visible = np.zeros((num_joints, 3), dtype=np.float32)
+
+                keypoints = np.array(obj['keypoints']).reshape(-1, 3)
+                joints_3d[:, :2] = keypoints[:, :2]
+                joints_3d_visible[:, :2] = np.minimum(1, keypoints[:, 2:3])
+
+                if 'center' in obj and 'scale' in obj:
+                    center = np.array(obj['center'])
+                    scale = np.array([obj['scale'], obj['scale']]) * 1.25
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_300w_dataset.py:86]: Give a comment for the meaning of `1.25`

File: mmpose/datasets/datasets/face/face_wflw_dataset.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        """Get normalize factor for evaluation.
+
+        Args:
+            gts (np.ndarray[N, K, 2]): Groundtruth keypoint location.
+
+        Return:
+            np.ndarray[N, 2]: normalized factor
+        """
+
+        interocular = np.linalg.norm(gts[:, 60, :] - gts[:, 72, :], axis=1)
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_wflw_dataset.py:119]: Explain `60` and `72`

Comment (innerlee) [reply mmpose/datasets/datasets/face/face_wflw_dataset.py:119]:
```python
interocular = np.linalg.norm(gts[:, 60, :] - gts[:, 72, :], axis=1, keepdims=True)
return np.tile(interocular, [1, 2])
```

File: mmpose/datasets/datasets/face/face_300w_dataset.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        masks = np.array(masks)
+
+        if 'NME' in metrics:
+            normalize_factor = self._get_normalize_factor(gts)
+            info_str.append(
+                ('NME', keypoint_nme(outputs, gts, masks, normalize_factor)))
+
+        return info_str
+
+    def evaluate(self, outputs, res_folder, metric='NME', **kwargs):
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_300w_dataset.py:156]: if `kwargs` is not used here, then add `assert len(kwargs) == 0`

File: mmpose/datasets/datasets/face/face_aflw_dataset.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+    `Annotated Facial Landmarks in the Wild: A Large-scale,
+    Real-world Database for Facial Landmark Localization.
+    In Proc. First IEEE International Workshop on Benchmarking
+    Facial Image Analysis Technologies, 2011`
+
+    The dataset loads raw images and apply specified transforms
+    to return a dict containing the image tensors and other information.
+
+    The landmark annotations follow the 68 points mark-up.
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_aflw_dataset.py:24]: Where to find the `68 points mark-up`? How about adding a docs page listing all keypoint formats?

File: mmpose/datasets/datasets/face/face_aflw_dataset.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                joints_3d = np.zeros((num_joints, 3), dtype=np.float32)
+                joints_3d_visible = np.zeros((num_joints, 3), dtype=np.float32)
+
+                keypoints = np.array(obj['keypoints']).reshape(-1, 3)
+                joints_3d[:, :2] = keypoints[:, :2]
+                joints_3d_visible[:, :2] = np.minimum(1, keypoints[:, 2:3])
+
+                if 'center' in obj and 'scale' in obj:
+                    center = np.array(obj['center'])
+                    scale = np.array([obj['scale'], obj['scale']]) * 1.25
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_aflw_dataset.py:82]: Explain `1.25`

File: mmpose/datasets/datasets/face/face_aflw_dataset.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                rec.append({
+                    'image_file': image_file,
+                    'center': center,
+                    'scale': scale,
+                    'rotation': 0,
+                    'joints_3d': joints_3d,
+                    'joints_3d_visible': joints_3d_visible,
+                    'dataset': self.dataset_name,
+                    'bbox': obj['bbox'],
+                    'box_size': obj['box_size'],  # normalize factor
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_aflw_dataset.py:97]: What does this comment ` # normalize factor` mean

File: mmpose/datasets/datasets/face/face_aflw_dataset.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+        Args:
+            res_file (str): Json file stored prediction results.
+            metrics (str | list[str]): Metric to be performed.
+                Options: 'NME'.
+
+        Returns:
+            dict: Evaluation results for evaluation metric.
+        """
+        info_str = []
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_aflw_dataset.py:126]: What if user pass an unsupported metric?

File: mmpose/datasets/datasets/face/face_aflw_dataset.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        box_sizes = np.array(box_sizes)
+
+        if 'NME' in metrics:
+            normalize_factor = self._get_normalize_factor(box_sizes)
+            info_str.append(
+                ('NME', keypoint_nme(outputs, gts, masks, normalize_factor)))
+
+        return info_str
+
+    def evaluate(self, outputs, res_folder, metric='NME', **kwargs):
```
Comment (innerlee) [inline mmpose/datasets/datasets/face/face_aflw_dataset.py:155]: also assert the `kwargs` to be empty here

File: mmpose/datasets/datasets/face/face_aflw_dataset.py (hunk 12 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:97]: I want to show that `box_size` is only used as a normalize factor for evaluation.

Comment (innerlee) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:97]: Okay, change it to `used as normalization factor`

File: mmpose/datasets/datasets/face/face_aflw_dataset.py (hunk 13 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:126]: This will not happen. There are checks, before calling this function.

File: mmpose/datasets/datasets/face/face_aflw_dataset.py (hunk 10 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:24]: It is hard to explain the keypoint formats in words. Images will be helpful to show the keypoint formats.

File: mmpose/datasets/datasets/face/face_aflw_dataset.py (hunk 13 above)
Comment (innerlee) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:126]: ok

File: mmpose/datasets/datasets/face/face_aflw_dataset.py (hunk 10 above)
Comment (innerlee) [reply mmpose/datasets/datasets/face/face_aflw_dataset.py:24]: Okay, any "official" link to this keypoint format?

**Comment by innerlee** [comment]
Reduce the img sizes if possible

image

File: tools/dataset/parse_cofw_dataset.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
+import json
```
Comment (innerlee) [inline tools/dataset/parse_cofw_dataset.py:1]: Any doc on how to use this script?

File: tools/dataset/parse_cofw_dataset.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+import os
+import time
+
+import cv2
+from hdf5storage import loadmat
```
Comment (innerlee) [inline tools/dataset/parse_cofw_dataset.py:6]: Can hdf5storage be replaced by h5py or scipy?

File: docs/data_preparation.md [hunk 17]
```diff
@@ -522,3 +526,179 @@ mmpose
...
+├── tests
+├── tools
+├── configs
+`── data
+    │── cofw
+        |── COFW_train_color.mat
+        |── COFW_test_color.mat
+```
+
+Run the following scripts:
```
Comment (innerlee) [inline docs/data_preparation.md:681]: Also mention the pwd location of the script

File: tools/dataset/parse_cofw_dataset.py (hunk 16 above)
Comment (jin-s13) [reply tools/dataset/parse_cofw_dataset.py]: `scipy.io` does not work for v7.3 or higher matlab.
I have replaced `hdf5storage` with `h5py`.

**Comment by innerlee** [comment]
Now the dataset docs need to be re-structured.

File: docs/tasks/2d_face_keypoint.md [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+- [300W](#300w-dataset) \[ [Homepage](https://ibug.doc.ic.ac.uk/resources/300-W/) \]
+- [WFLW](#wflw-dataset) \[ [Homepage](https://wywu.github.io/projects/LAB/WFLW.html) \]
+- [AFLW](#aflw-dataset) \[ [Homepage](https://www.tugraz.at/institute/icg/research/team-bischof/lrs/downloads/aflw/) \]
+- [COFW](#cofw-dataset) \[ [Homepage](http://www.vision.caltech.edu/xpburgos/ICCV13/) \]
+
+## 300W Dataset
+
+For 300W data, please download images from [300W Dataset](https://ibug.doc.ic.ac.uk/resources/300-W/).
+Please download the annotation files from [300w_annotations](https://download.openmmlab.com/mmpose/datasets/300w_annotations.tar).
+Extract them under {MMPose}/data, and make them look like this:
```
Comment (innerlee) [inline docs/tasks/2d_face_keypoint.md:17]: `{MMPose}/data`

File: mmpose/models/keypoint_heads/top_down_simple_head.py [hunk 19]
```diff
@@ -79,12 +98,78 @@ def __init__(self,
...
         x = self.final_layer(x)
         return x
 
+    def _init_inputs(self, in_channels, in_index, input_transform):
+        """Check and initialize input transforms.
+
+        The in_channels, in_index and input_transform must match.
+        Specifically, when input_transform is None, only single feature map
+        will be selected. So in_channels and in_index must be of type int.
+        When input_transform
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_simple_head.py:112]: incomplete sentence

File: mmpose/models/keypoint_heads/top_down_simple_head.py [hunk 20]
```diff
@@ -104,7 +104,7 @@ def __init__(self,
                             out_channels=conv_channels,
                             kernel_size=num_conv_kernels[i],
                             stride=1,
-                            padding=padding))
+                            padding=1 if num_conv_kernels[i] == 3 else 0))
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_simple_head.py:107]: What's the philosophy of choosing padding? Should it be `(kernelsize - 1) // 2`?

File: mmpose/models/utils/ops.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Upsample(nn.Module):
+
+    def __init__(self,
+                 size=None,
+                 scale_factor=None,
+                 mode='nearest',
+                 align_corners=None):
+        super(Upsample, self).__init__()
```
Comment (innerlee) [inline mmpose/models/utils/ops.py:39]: super

File: mmpose/models/utils/ops.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+                        f'When align_corners={align_corners}, '
+                        'the output would more aligned if '
+                        f'input size {(input_h, input_w)} is `x+1` and '
+                        f'out size {(output_h, output_w)} is `nx+1`')
+    if isinstance(size, torch.Size):
+        size = tuple(int(x) for x in size)
+    return F.interpolate(input, size, scale_factor, mode, align_corners)
+
+
+class Upsample(nn.Module):
```
Comment (innerlee) [inline mmpose/models/utils/ops.py:32]: The name conflicts with pytorch's https://pytorch.org/docs/stable/generated/torch.nn.Upsample.html

**Comment by innerlee** [comment]
Some  minor comments

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
Need to update several readmes, docs, merge_doc/stats scripts, etc.
